pkg/adaptor/controller.go

```diff
@@ -0,0 +1,365 @@
+	namespace, name, service, err := splitTrafficSplitKey(key)
+	if err != nil {
+		log.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the Ts resource with this namespace/name
+	ts, err := c.tsclientset.SplitV1alpha1().TrafficSplits(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+
+		// TS does not exist anymore
+		// Check if there is a relevant SP that was created or updated by SMI Controller
+		// and clean up its dstOverrides
+		log.Infof("trafficsplit/%s is deleted, trying to cleanup the relevant serviceprofile", name)
+		sp, err := c.spclientset.LinkerdV1alpha2().ServiceProfiles(namespace).Get(ctx, c.toFQDN(service, namespace), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if ignoreAnnotationPresent(sp) {
+			log.Infof("skipping clean up of serviceprofile/%s as ignore annotation is present", sp.Name)
+			return nil
+		}
+
+		// Empty dstOverrides in the SP
+		sp.Spec.DstOverrides = nil
+		_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(namespace).Update(ctx, sp, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+		log.Infof("cleaned up `dstOverrides` of serviceprofile/%s", sp.Name)
+		return nil
+	}
+
+	serviceFQDN := c.toFQDN(ts.Spec.Service, ts.Namespace)
+	// Check if the Service Profile is already present
+	sp, err := c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Get(ctx, serviceFQDN, metav1.GetOptions{})
+	if err != nil {
+		// Return if its not a not found error
+		if !errors.IsNotFound(err) {
+			log.Errorf("couldn't retrieve serviceprofile/%s: %s", serviceFQDN, err)
+			return err
+		}
+
+		// Create a Service Profile resource as it does not exist
+		sp, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Create(ctx, c.toServiceProfile(ts), metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		log.Infof("created serviceprofile/%s for trafficsplit/%s", sp.Name, ts.Name)
+	} else {
+		log.Infof("serviceprofile/%s already present", sp.Name)
+		if ignoreAnnotationPresent(sp) {
+			log.Infof("skipping update of serviceprofile/%s as ignore annotation is present", sp.Name)
+			return nil
+		}
+
+		// Check if SP Matches the TS, and update if it not
+		spFromTs := c.toServiceProfile(ts)
+		if !equal(spFromTs, sp) {
+			log.Infof("serviceprofile/%s does not match trafficscplit/%s", sp.Name, ts.Name)
+			updateDstOverrides(sp, ts, c.clusterDomain)
+			_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Update(ctx, sp, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+			log.Infof("updated serviceprofile/%s as it's not equivalent to trafficsplit/%s", sp.Name, ts.Name)
+		}
+	}
+
+	return nil
+}
+
+func ignoreAnnotationPresent(sp *serviceprofile.ServiceProfile) bool {
+	_, ok := sp.Annotations[ignoreServiceProfileAnnotation]
+	return ok
+}
+
+func equal(spA *serviceprofile.ServiceProfile, spB *serviceprofile.ServiceProfile) bool {
```

Comment:
I'll echo what I said on the other PR: I don't think that there's a significant cost to doing redundant updates so I'd be inclined to skip the equals check.